[API-1530] Update Compact serialization documentation

mdumandag opened the pull request:
This PR updates the Compact serialization documentation by
- making it more polyglot friendly
- adding frequently requested sections including information for
	- zero-config interoperability
	- WAN support
	- links to SQL mappings
- fully describing the types supported

mdumandag committed:
Update Compact serialization documentation

This PR updates the Compact serialization documentation by
- making it more polyglot friendly
- adding frequently requested sections including information for
	- zero-config interoperability
	- WAN support
	- links to SQL mappings
- fully describing the types supported

JakeSCahill commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -326,77 +342,1227 @@ hazelcast:
+When a class is serialized using the zero-config Compact serializer, Hazelcast will choose the
+fully qualified class name for Java as the type name automatically.
+
+Below is the way to register an explicit serializer for a certain class.
```
```suggestion
To register an explicit serializer for a certain class, add the following configuration:
```

JakeSCahill commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -326,77 +342,1227 @@ hazelcast:
+value types. The names of those arrays are of the form `fieldName + '!keys'` and
+`fieldName + '!values'`.
+
+For Java, zero-config Compact serializer uses reflection to read and write to fields of
```
```suggestion
For Java, a zero-config Compact serializer uses reflection to read and write to fields of
```

JakeSCahill commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -326,77 +342,1227 @@ hazelcast:
+The same holds true for the Java records. Hazelcast supports serializing and deserializing
+Java records, without an extra configuration as well.
+
+Assuming the `Employee` class above were a Java record:
```
```suggestion
Assuming the `Employee` class was a Java record:
```

mdumandag commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -82,41 +77,66 @@ public class Employee {
+.NET::
++
+--
+[source,cs]
```
Yes, I will coordinate the effort for adding these once the PR is merged.

mdumandag committed:
address review comments

mdumandag commented:
Thanks a lot for the review @JakeSCahill. I addressed your comments in https://github.com/hazelcast/hz-docs/pull/554/commits/09c94e5351796f10028f4cec5f63c0c8cbad2ec7

srknzl commented:
Note: I won't review grammar related stuff assuming @JakeSCahill already checked them.

srknzl commented:
IMO, the configuration section can be in an upper place than its current place:

![image](https://user-images.githubusercontent.com/32355782/200424652-b63519af-3e09-45b3-aac2-b6fd47242662.png)

We show how to configure a compact serializer at the end of the first section, Implementing CompactSerializer without much explaining it. I think we can put the configuration section right after the first one.

srknzl commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -326,77 +342,1221 @@ hazelcast:
+Lastly, the following is a sample configuration that registers zero-config
+serializer for a certain class, without implementing an explicit serializer.
+
+This way, one can override other serialization mechanisms for a certain class
+such as Java serializable, using zero-config serializer.
```
```suggestion
This way, one can override the serializer of a certain class, such as the Java `Serializable` serializer, 
with the zero-config serializer.
```

srknzl commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -326,77 +342,1221 @@ hazelcast:
+== SQL Support
+
+Compact serialized objects can be used in SQL statements, provided that a mappings are created,
+similar to other serialization formats. See xref:sql:mapping-to-maps.adoc[], to learn more
```
maybe give a link to the `#compact-objects` section

srknzl commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -326,77 +342,1221 @@ hazelcast:
+
+However, since the Compact serialization is promoted to the stable status in 5.2, and it is not
+compatible with the previous BETA versions, one has to make sure that the whole WAN cluster
+topology, including all senders and receivers, is at least as new as 5.2, before start replicating data structures containing Compact
```
```suggestion
topology, including all senders and receivers, are at least as new as 5.2, before starting to replicate data structures containing Compact
```
I was not going to check grammar but, am I missing something, should not this be plural ? Also, it should be `before starting to replicate`

mdumandag replied:
I was referring to `topology`, but changed it to `topology members` and made it plural

srknzl commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -326,77 +342,1221 @@ hazelcast:
+topology, including all senders and receivers, is at least as new as 5.2, before start replicating data structures containing Compact
+serialized objects.
+
+Since it has promoted to the stable status, it will be possible in the future releases to
+replicate Compact serialized objects between different WAN clusters of versions at least as
+new as 5.2.
```
```suggestion
Since it has promoted to the stable status, replicating Compact serialized objects between WAN clusters are of different versions will be possible in the future releases provided that all clusters' versions are at least 5.2.
```

Also I think `at least as new as 5.2` feels long. If you change it, don't forget to change other occurrences.

mdumandag replied:
I have removed it from here, but left it in the first paragraph, as I believe it was necessary there

mdumandag commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -35,33 +35,28 @@ a class to implement a particular interface. Serializers might be implemented an
 
-== Using Compact Serialization With Zero-Configuration
+NOTE: Compact serialization is promoted to the stable status in the 5.2 Hazelcast release.
+The older versions released under the BETA status are not compatible with stable the 5.2 version. +
```
This is to create a line break and start the next sentence in a new paragraph

mdumandag commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -326,77 +342,1221 @@ hazelcast:
+* `Set` or `HashSet` of the types listed above and first class types, represented by their
+respective arrays with the same field name.
+* `Map` or `HashMap` or the types listed above and first class types, represented by two arrays
+per field, one for keys and one for values, represented by their respective arrays for key and
```
Makes sense. I will remove it

mdumandag replied:
I meant for each Map field of class, we will have two fields in the binary

mdumandag commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -326,77 +342,1221 @@ hazelcast:
+enable reflective access for packages of your module by using the `opens` statement in the
+module declaration:
+
+[source,text]
```
It is not following the Java syntax, and the renderer (at least in my IDE) does not colorize it if I change that to Java. So, I plan to keep it as it is

mdumandag commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -326,77 +342,1221 @@ hazelcast:
+| Array of timestamp with timezones or null.
+
+| COMPACT
+| Can be any type.
```
I changed it to `Can be any user type`. Is it OK?

mdumandag committed:
address review comments

mdumandag commented:
@srknzl I have addressed your comments, and as you suggested moved the config section up. Also, I have removed the duplicate configuration snippets from the `Implementing CompactSerializer` section by simply adding a reference to the configuration section.

I have also added some notes about the performance aspect of the zero-config serialization

srknzl reviewed (approved):
lgtm, good job 👏🏻

Serdaro commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -35,368 +35,1390 @@ a class to implement a particular interface. Serializers might be implemented an
+The older versions released under the BETA status are not compatible with stable the 5.2 version. +
+For future versions, the backward compatibility will be maintained, just like any other
+Hazelcast feature.
 
```
@mdumandag , don't you think this a release notes or migration guide material? (because our product documentation are versioned)

Serdaro replied:
Maybe we can reword the related entry at https://docs.hazelcast.com/hazelcast/5.2/release-notes/5-2-0#breaking-changes

mdumandag replied:
Yes, we can put this information in a couple of places, but my impression was that since we were beta in 5.1 and 5.0 for compact, and only promoted to the stable status in 5.2, I don't think it is a breaking change. We never gave a guarantee of future compatibility for compact beta

mdumandag replied:
But, this is definitely important information, so maybe we can put it in some place in the release notes

OzanCansel committed:
Update compact-serialization.adoc

OzanCansel committed:
Update compact-serialization.adoc

harunalpak committed:
Update nodejs part of compact-serialization.adoc

mdumandag commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -35,368 +35,1501 @@ a class to implement a particular interface. Serializers might be implemented an
+--
+[source,javascript]
+----
+Client.newHazelcastClient(
```
@srknzl @harunalpak I assume this is a Javascript code, not TypeScript. Is it OK for the code samples? Same for the other Node.js code snippets

srknzl replied:
IMO, using javascript in code samples is better. Using our client with typescript is just an option. We use javascript in our code samples as well. Typescript users can easily convert these examples to TS.

mdumandag commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -35,368 +35,1501 @@ a class to implement a particular interface. Serializers might be implemented an
+
+[cols="1m,1a,1a,1a,1a,1a,1a,1a]
+|===
+| Type | Java | .NET | C++ | Node.js | Python | Go | Description
```
@harunalpak can you also please fill this table with the type mappings for Node.js?

harunalpak replied:
Fixed.

mdumandag commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -35,368 +35,1501 @@ a class to implement a particular interface. Serializers might be implemented an
+{
+    int64_t id;
+    std::string name;
+    int32_t age;
```
@OzanCansel can we please add the comments present in Java and Python code snippets here, to make them consistent? Note that, there are a few other code snippets down below, which lacks comments in C++

mdumandag commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -35,368 +35,1501 @@ a class to implement a particular interface. Serializers might be implemented an
+class Employee {
+    id;
+    name;
+    age;
```
@harunalpak  can we please add the comments present in Java and Python code snippets here, to make them consistent? Note that, there are a few other code snippets down below, which lack comments in Node.js

harunalpak replied:
I added comments.

mdumandag commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -35,368 +35,1501 @@ a class to implement a particular interface. Serializers might be implemented an
+|
+| Array of timestamps or null.
+
+| TIMESTAMP_WITH_TIMEZONE
```
@OzanCansel I see that we didn't put anything to this and `ARRAY_OF_TIMESTAMP_WITH_TIMEZONE ` column for C++. Is there a reason for that?

Commit:
some fixes related to PR comments

devOpsHazelcast commented:
[![CLA assistant check](https://cla.hazelcast.org:443/pull/badge/not_signed)](https://cla.hazelcast.org:443/hazelcast/hz-docs?pullRequest=554) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.hazelcast.org:443/hazelcast/hz-docs?pullRequest=554) before we can accept your contribution.<br/>**5** out of **6** committers have signed the CLA.<br/><br/>:white_check_mark: mdumandag<br/>:white_check_mark: OzanCansel<br/>:white_check_mark: harunalpak<br/>:white_check_mark: zpqrtbnk<br/>:white_check_mark: yuce<br/>:x: Harun<br/><hr/>**Harun** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.hazelcast.org:443/check/hazelcast/hz-docs?pullRequest=554) it.</sub>

mdumandag commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -35,368 +35,1506 @@ a class to implement a particular interface. Serializers might be implemented an
+| Boolean
+| 
+| boost::optional<bool>
+| boolean
```
@harunalpak I think this should be `boolean | null` for Node.js. The same applies to others as well. Can we make sure that the type mappings here are the same as our API at here https://github.com/hazelcast/hazelcast-nodejs-client/blob/master/src/serialization/compact/CompactWriter.ts?

harunalpak replied:
Fixed

Commit:
some fixes related to PR comments

Commit:
some fixes related to PR comments

zpqrtbnk committed:
Implement .NET examples (#1)

mdumandag committed:
fix formatting issues for the type mappings table

zpqrtbnk committed:
More .NET Stuff (#2)

* Implement .NET examples

* Implement .NET types

yuce committed:
Updated for Go client (#3)

* Updated for Go client

* fix string

OzanCansel committed:
Filled columns for missing types. Comments added.

OzanCansel committed:
Update compact-serialization.adoc

mdumandag commented:
@Serdaro we are ready to merge this PR. Will you review it as well, or the review from Jake is enough? Also, we would like to backport this to 5.2.0 and 5.2.1. How can we do that?

Serdaro commented:
@mdumandag , Let me do a quick review.
It will be enough to put the "backport to 5.1" and "backport to 5.2" labels to the PR before merge.

Serdaro commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -35,368 +35,1620 @@ a class to implement a particular interface. Serializers might be implemented an
-serializing and deserializing instances of that class. If the automatic schema
-extraction fails, Hazelcast throws an exception.
+NOTE: Compact serialization is promoted to the stable status in the 5.2 Hazelcast release.
+The older versions released under the BETA status are not compatible with stable the 5.2 version. +
```
... with the stable 5.2 version.

srknzl commented on docs/modules/serialization/pages/compact-serialization.adoc:
```diff
@@ -35,368 +35,1620 @@ a class to implement a particular interface. Serializers might be implemented an
+
+== SQL Support
+
+Compact serialized objects can be used in SQL statements, provided that a mappings are created,
```
```suggestion
Compact serialized objects can be used in SQL statements, provided that a mapping is created,
```

mdumandag replied:
Changed it to `provided that mappings are created`

mdumandag committed:
address review comments
